### PR TITLE
Interface to older style Optimizable from new graph based Optimizable

### PR DESCRIPTION
**Pull request by mbkumar**
Graph based Optimizable class now allows for external handling of dofs. This change in behavior allows derivative classes to implement their own methods to handle the dofs data, which is the behavior of the subclasses of the current(older style) Optimizable class.

#### Initialization
In addition to the previous arguments allowed with `_core.graph_optimizable.Optimizable` class, one more argument, `external_dof_setter` is implemented. This enables the graph based Optimizable class to keep track of the dofs data by storing it in the `Dofs` class, but at the same time, a copy of the dofs data is handled by the `external_dof_setter` method. This way, the current derivative classes of older `Optimizable` can switch to graph based Optimizable class without major design changes.

#### local dof setter
Since the handling of the dependencies between Optimizable objects is carried out by the base `_core.graph_optimizable.Optimizable` class, only the `local_x` setter needs to be modified to be consistent with the above design goal. So the  `local_x.setter` after storing the data in the Dofs object passes `local_full_x` (which is nothing but full x data stored in Dofs object) to `external_dof_setter` supplied. As a result, derivative class developers don't have to worry about writing their own code to change dofs status from fixed to free or vice versa and to handle dependencies between different Optimizable objects.

#### Example class
Lets take a subclass of older style Optimizable class, which implements set_dofs and get_dofs methods.
```python
class SubOptimizable(simsopt._core.optimizable.Optimizable):
      def __init__(self):
            ...
            Optimizable.__init__(self, **kwargs)
      def get_dofs(self):
            ....
      def set_dofs(self, x):
            ....
```
The new code is 
```python
class SubOptimizable(simsopt._core.graph_optimizable.Optimizable):
      def __init__(self):
            ...
            Optimizable(self, external_dof_setter=SubOptimizable.set_dofs, x0=self.get_dofs(), **kwargs)
      def get_dofs(self):
            ....
      def set_dofs(self, x):
            ....
```
The new init code will differ even more than shown here to get rid of custom handling of dependencies and other things and offload them to the new graph based Optimizable class.

### CODE REVIEW

**Comment by mbkumar** [comment]
@florianwechsung 

I implemented a new subclass of graph based Optimizabled that allows for set_dofs and get_dofs. Please take a look at the new class and we can work together to convert some of the existing classes that use old style Optimizable base class to the newer graph based approach.

@landreman,
We could simultaneously try to refactor the MHD classes to use the new framework.

**Comment by mbkumar** [comment]
Curve classes are modified to conform with the graph framework.

**Comment by florianwechsung** [comment]
Two quick comments before I look closer at this:
- `PandasDOFs` has zero coverage, so i assume it's not used anymore? Why not delete it?
- same for `oldgraph_surface.py`

**Comment by mbkumar** [comment]
I am keeping it as a placeholder for now, because there is some push on
pandas to make indexing faster. If it materializes, we can go back to
pandas. I am going to reuse oldgraph_surface.py in the new branch I am
working on for surfaces.

As a workaround, I'll create another branch to keep these files and delete
the 2 files from the external_dof_handling branch.


Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)

**Comment by mbkumar** [comment]
Removed oldgraph_surface.py and pandas DOFs class

Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)

**Review by florianwechsung — DISMISSED** [review]

**Review by landreman — COMMENTED** [review]
We had also talked about renaming `opts_in` -> `parents`, which I think is more descriptive and specific.

File: src/simsopt/geo/curve.py [hunk 1]
```diff
@@ -1,12 +1,24 @@
 from math import sin, cos
+from abc import abstractmethod
```
[the remark this answers is not in the document]
Comment (mbkumar) [reply src/simsopt/geo/curve.py:2]: You are right. I removed it.

File: src/simsopt/geo/curvexyzfourier.py [hunk 2]
```diff
@@ -32,7 +31,8 @@ def __init__(self, quadpoints, order):
         elif isinstance(quadpoints, np.ndarray):
             quadpoints = list(quadpoints)
         sopp.CurveXYZFourier.__init__(self, quadpoints, order)
-        Curve.__init__(self)
+        Curve.__init__(self, x0=self.get_dofs(),
```
[the remark this answers is not in the document]
Comment (mbkumar) [reply src/simsopt/geo/curvexyzfourier.py:35]: @florianwechsung, are you in agreement with Matt's scheme, or do you have any other naming convention in mind?

[the remark this answers is not in the document]
Comment (mbkumar) [reply src/simsopt/geo/curvexyzfourier.py]: Taken care of.

**Comment by mbkumar** [comment]
> We had also talked about renaming `opts_in` -> `parents`, which I think is more descriptive and specific.

Done.

**Comment by mbkumar** [comment]
@landreman @florianwechsung, If you don't have any additional comments, I'll merge this PR.

**Review by landreman — APPROVED** [review]

**Pull request merged**